### ISSUE TITLE
Improve Markdown syntax highlight for lists and headings

### ISSUE THREAD

**Issue by Stakker** [microsoft/vscode-markdown-tm-grammar#92]
According to the CommonMark spec, you should not have to add a line break before a list. I personally also fully support this, as having those extra empty lines in the middle of text wastes a lot of vertical space and hurts readability. No line break feels natural and is generally supported by Markdown implementations.

This is the CommonMark description:
https://spec.commonmark.org/0.30/#principle-of-uniformity
(The description about lists interrupting paragraphs starts a few lines above.)

In short, I want to be able to write lists like this:

    This is a list
    - List item
    - Another list item

Not like this:

    This is a list

    - List item
    - Another list item

Cheers!

**Comment by mjbvz** [comment]
I've moved the issue to the repository that contains the markdown grammar used for syntax highlighting. Please take a look at updating it to fix this issue

**Comment by dev-manuel** [comment]
So if I understood that right the first part of this issue is about the formatter; if you format the document it will add a line break between heading and entries of the list as well it adds indention to the entries and you want both of them to be changed.
The second would then be about the highlighting (coloring in this case) which is done automatically. Am I correct with that assumption? 
If I am right with that, my question for @mjbvz is the following: Is this repository also responsible for the formatter part or just for the grammar (which includes the coloring part)? I experienced that the markdown formatter is not part of a fresh build of vscode. The formatter you @Stakker are using is likely provided by an external extension e.g. Prettier?

**Comment by dev-manuel** [comment]
Also why is there a file extensions/markdown-basics/syntaxes/markdown.tmLanguage.json in the vscode repo that also contains a grammar for markdown (they are slightly different)? What's the difference between these two?

**Comment by Stakker** [comment]
I am not using a formatter, this is all about syntax highlight/colors (which uses TextMate grammar, I think?). Let's forget the line break for now — that was just a misunderstanding of what is happening from me :-)

So, it seems that the only problem is that not all of Markdown syntax is implemented in the grammar (or whatever you officially call it... VS Code only highlights some Markdown syntax).

Headlines (using '#') and code blocks are highlighted. Lists are not highlighted at all. At least that's what it looks like.

So, what needs to be done is implement lists into the Markdown grammar. (Start a new line with a dash (minus), plus or an asterisk or a number follower by a period, and it is recognized as a list and colored accordingly. And you should not need to add a line break before a list for it to be recognized as a list.)

If you are going to be poking in the grammar, it would also be great it could also recognize headings underlined with "-----" or "=====", because now the syntax only understands headings that start with '#'.

There is also more (https://www.markdownguide.org/basic-syntax), I think these are the most essential basics.

**Comment by dev-manuel** [comment]
Allright, implementing the underline to make it a header should be doable.
But how do you like lists to be highlighted? If I run VSCode without other extensions (so just the vscode-markdown-tm-grammar is enabled) at least the dashes (or enumerations like "1.") of the list entries are already colored blue-ish instead of white (stock dark theme).

**Comment by Stakker** [comment]
Ahh, so it does... It seems I'm using a theme (one of the Dayle Rees themes) that doesn't support it.

I did check it out with the default theme earlier, but then I was still misusing the code block syntax thinking they are lists, so I missed that. It also seems that the default theme doesn't implement code block syntax, but it is in the grammar, since the Dayle Rees theme highlights it.

I definitely think just highlighting the first character (especially a tiny dash) is not as clear as highlighting the whole contents of the list though. Attached is an example of how Sublime Text does it, and I think this is just much more useful and clear.

I would also like for the underlined headlines to be highlighted, not just the underline.

But now I'm not sure how much the grammar can change this behavior? Is this actually up to the theme or the grammar to decide how much is highlighted? Sorry about my ignorance here, I don't really know anything about the innards of VS Code.

![Sublime](https://user-images.githubusercontent.com/8972595/135769825-33f8607e-a5d5-46ff-a12a-c6d366a52b1d.png)

**Comment by dev-manuel** [comment]
Okay, I will work on the implementation of entries that are completely colored and headings that are defined with underlinings (with '#' and '-').

**Comment by Stakker** [comment]
I appreciate you looking into this!

**Comment by mjbvz** [comment]
@dev-manuel Yes thanks for investigating. VS Code has never provided a built-in markdown formatter so this issue is focused just on the syntax highlighting part of the issue

Also keep in mind that how code is colored is are ultimately up the themes. We just want to make sure the parsed document has the right set of tokens so that themes can color it as they wish

**Comment by mjbvz** [comment]
Also looking over the original issue again, both style lists appear to work for me:

<img width="817" alt="Screen Shot 2021-10-04 at 9 34 19 AM" src="https://user-images.githubusercontent.com/12821956/135889574-3bdb9c4a-f37a-4706-948c-ea1fe6f444b6.png">

@Stakker Have you tested with other themes to confirm this isn't a theme specific problem?

**Comment by Stakker** [comment]
Ah, so it *is* up to the theme after all... I have tried other themes, but I managed to the pick exactly the ones that don't fully color lists :-P (Like Monokai, and also the default theme.)

Thank you for your patience and sorry about wasting your time with this... At least I learned something :-) I guess I just need to learn how to customize/make themes.

**Comment by mjbvz** [comment]
Ok, thanks for following up. Closing as by-design

**Comment by Stakker** [comment]
The best option I think would be for the grammar to support coloring the heading text, and then the theme can decide whether to color just the underline or the whole shebang. (Identical to the list syntax.)
